Objective-C/CBLDatabase.mm

```diff
@@ -360,58 +360,83 @@ - (BOOL) inBatch: (NSError**)outError usingBlock: (void (NS_NOESCAPE ^)())block
 #pragma mark - DATABASE MAINTENANCE
 
 - (BOOL) close: (NSError**)outError {
+    NSArray *activeReplicators, *activeLiveQueries = nil;
+    
     CBL_LOCK(self) {
-        if (_c4db == nullptr)
+        if ([self isClosed] || _isClosing)
             return YES;
         
         CBLLogInfo(Database, @"Closing %@ at path %@", self, self.path);
-    
-        if (_activeReplications.count > 0) {
-            return createError(CBLErrorBusy, kCBLErrorMessageCloseDBFailedReplications, outError);
-        }
         
-        if (_liveQueries.count > 0) {
-            return createError(CBLErrorBusy, kCBLErrorMessageCloseDBFailedQueryListeners, outError);
-        }
+        _isClosing = YES;
         
-        [self cancelDocExpiryTimer];
+        _closeCondition = [[NSCondition alloc] init];
+        
+        activeReplicators = [_activeReplicators allObjects];
+        
+        activeLiveQueries = [_activeLiveQueries allObjects];
+    }
     
-        C4Error err;
-        if (!c4db_close(_c4db, &err))
-            return convertError(err, outError);
+    // Stop active replicators:
+    for (CBLReplicator* r in activeReplicators) {
+        [r stop];
+    }
     
-        [self freeC4Observer];
-        [self freeC4DB];
+    // Stop active live queries:
+    for (CBLLiveQuery* q in activeLiveQueries) {
+        [q stop];
+    }
     
-        return YES;
+    // Wait for all active replicators and live queries to stop:
+    [_closeCondition lock];
+    while (![self isReadyToClose]) {
+        [_closeCondition wait];
     }
-}
-
-- (BOOL) delete: (NSError**)outError {
+    [_closeCondition unlock];
+    
     CBL_LOCK(self) {
-        [self mustBeOpen];
-        
-        if (_activeReplications.count > 0) {
-            return createError(CBLErrorBusy, kCBLErrorMessageDeleteDBFailedReplications, outError);
-        }
-        
-        if (_liveQueries.count > 0) {
-            return createError(CBLErrorBusy, kCBLErrorMessageDeleteDBFailedQueryListeners, outError);
-        }
-        
+        // Cancel doc expiry timer:
         [self cancelDocExpiryTimer];
         
+        // Free C4Observer:
+        [self freeC4Observer];
+        
+        // Close database:
+        BOOL success = YES;
         C4Error err;
-        if (!c4db_delete(_c4db, &err))
-            return convertError(err, outError);
+        if (!c4db_close(_c4db, &err))
+            success = convertError(err, outError);
         
-        [self freeC4Observer];
-        [self freeC4DB];
+        // Release database:
+        if (success)
+            [self freeC4DB];
         
-        return YES;
+        // Reset closing flag:
+        _isClosing = NO;
```

Comment:
All these flags... Would it be better to have a state machine?
CLOSED, OPEN, CLOSING.  I think that's sufficient...

Reply:
It created https://issues.couchbase.com/browse/CBL-794 to evaluate and track this.